xc/common/utils/prjxray_create_ioplace.py

```diff
@@ -144,6 +164,21 @@ def main():
             sys.exit(1)
 
         loc, is_output, iob = pad_map[pad]
+
+        if args.synth_tiles:
```

Comment:
Something that is confusing me now is that pad_map contains an entry for the pad, so why is this if statement needed?  Is the data from pad_map wrong?  Why not just fix that data instead of adding the code here?  Feels inconsistent to me.  Thoughts?

Reply:
So my thought is that for something like an overlay or an roi that aligns with a complex overlay, the format of the pad map doesn't make a whole lot of sense because the pad map csv maps the synth_tile to some real IO.  I have been using names like SYN0 to represent a synth pad that doesn't directly correspond to a real pad.

In an roi region it doesn't matter whether you use information directly from the pad map or from the synth_tiles, which is why it's not currently enabled. But, for overlay generation we would either have to reconsider the pad map csv format to handle synth IOs that don't directly correspond to a real IO and change the pin map csv generation to handle both synth ios and real ios at the same time, or just take the synth tiles as an input to create_ioplace.

This method was the easiest to get overlay generation working, but maybe the better solution is to figure out the correct format for the pinmap csv and merge together the synth tile to pinmap csv with the normal pinmap csv script.

Another method that could make this more consistent, is completely remove the pinmap csv generation for synth_tiles and only use the args.synth_tiles option in create_ioplace.

Thoughts?

Reply:
> But, for overlay generation we would either have to reconsider the pad map csv format to handle synth IOs that don't directly correspond to a real IO and change the pin map csv generation to handle both synth ios and real ios at the same time, or just take the synth tiles as an input to create_ioplace.

I'd rather have the relevant code to generate the correct pin map csv rather than the current code?  Pinmap CSV is simply a list of pads to VPR tile locations.  In the case where they are purely synthetic, the name is invented, but still has meaning.  Like eventually you'll have `CLK_MHZ` or `BUS_READY` or something, yes?

Reply:
Something that is bothering me about this change is that it isn't being tested at all right now, because that code path is unused.  It feels weird to have dead code being added.

Reply:
> > But, for overlay generation we would either have to reconsider the pad map csv format to handle synth IOs that don't directly correspond to a real IO and change the pin map csv generation to handle both synth ios and real ios at the same time, or just take the synth tiles as an input to create_ioplace.
> 
> I'd rather have the relevant code to generate the correct pin map csv rather than the current code? Pinmap CSV is simply a list of pads to VPR tile locations. In the case where they are purely synthetic, the name is invented, but still has meaning. Like eventually you'll have `CLK_MHZ` or `BUS_READY` or something, yes?

Yes.  I guess what I mean by "the format of the pad map doesn't make sense" is what should go in the iob column for synth tiles that do not correspond to a real IO?

Current pinmap format:
```
name,x,y,z,is_clock,is_input,is_output,iob
A1,114,73,0,1,1,1,IOB_X1Y81
```

> Something that is bothering me about this change is that it isn't being tested at all right now, because that code path is unused. It feels weird to have dead code being added.

I agree completely now that I'm thinking about it.  This shouldn't be merged as is.  Maybe the correct thing to do it delete the unused code in this pr and include the merged pinmap generation fixes in the overlay generation pr?

Reply:
> I agree completely now that I'm thinking about it. This shouldn't be merged as is. Maybe the correct thing to do it delete the unused code in this pr and include the merged pinmap generation fixes in the overlay generation pr?

That sounds good to me.  I think a focused PR on dealing with overlay generation issues might be more productive, as the other changes in this PR all otherwise look good and generalize well.

Reply:
> That sounds good to me. I think a focused PR on dealing with overlay generation issues might be more productive, as the other changes in this PR all otherwise look good and generalize well.

Now that I'm thinking about it more, this fix is necessary for heterogeneous synth tiles to work.  Currently the pinmap from synth tiles always assigns the z value to 0, so if there are multiple synth tiles at the same location the pinmap will be wrong.  It doesn't get tested currently because there is no test in here that actually uses multiple synth tiles at the same location yet (that plan was to add that with the overlay generation test).

Reply:
Ok, given that, shouldn't we update the CMake to pass the synth tiles JSON?  Or fix the tool to create the pinmap CSV so that it works?

Reply:
> Ok, given that, shouldn't we update the CMake to pass the synth tiles JSON? Or fix the tool to create the pinmap CSV so that it works?

I think I should fix the tool to create the pinmap csv.  If I merge together the synth_tile_to_pinmap_csv and create_pinmap_csv, can I just put some arbitrary name in the iob column for synth_tiles whose pad name isn't in the real io pad list?  Should I add another column to the csv for whether it's an non-real-io-associated synth_tile?  I think the iob is only used for setting io standards, so I just need to not set those for the unassociated synth_tiles.

Reply:
> can I just put some arbitrary name in the iob column for synth_tiles whose pad name isn't in the real io pad list

I'd just leave it blank

> Should I add another column to the csv for whether it's an non-real-io-associated synth_tile

I think this is prudent